demo/transient/transient_snes_mpfaof90.F90

```diff
@@ -428,14 +443,19 @@ program main
     call TDyPostSolveSNESSolver(tdy,U,ierr);
     CHKERRA(ierr);
 
-    ! TODO: Mass conservation diagnostic output is disabled for the moment.
+    ! Check the total liquid mass after the step. Because the liquid mass is
+    ! nonnegative, the sum over all cells is the same as the 1-norm.
+    call TDyComputeDiagnostics(tdy, diags_dm, diags, ierr)
```

Comment:
1. `TDyComputeDiagnostics` is getting a strided `Vec` with all diagnostics. So, how about renaming it to `TDyGetDiagnostics`?
2. Apart from renaming as suggested above, how about also doing the following?
   - Create another function named `TDyComputeDiagnostics` that fills all the diagnostics in a strided `Vec diags`, where `diags` is owned by TDycore.
   - Then, a function like `TDyGetSaturation` can use `VecStrideGather` to extract the subset from `diags`?

Reply:
`TDyComputeDiagnostics` is actually doing what your suggested `TDyComputeDiagnostics` function would do. It doesn't allocate or get the `diags` vector--that is done on line 385, using the DM created by `TDyCreateDiagnostics`. `TDyComputeDiagnostics` simply computes and fills in the strided diagnostics vector.

Regarding `TDyGetSaturation`: Jed says we have to create a "sub-DM" from `diags_dm` that can create a vector that stores only the saturation vector. That would have to be done somewhere in order to get a vector that can hold the saturation (to be passed to `TDyGetSaturation`), so we're looking at another function to create a sub vector (from the sub-DM) for each diagnostic. It seems a little cumbersome to me to have all these DMs floating around, but maybe I'm just not used to manipulating function spaces in the same way as we manipulate fields.

To summarize, we currently have:

1. `TDyCreateDiagnostics`: creates a DM that manages the layout for all diagnostic fields (from which a strided diagnostics vector can be created using `DMCreateGlobalVector`).
2. `TDyComputeDiagnostics`: given the diagnostics DM and a diagnostics vector, populates the vector with the updated diagnostics fields.

I think (in your first comment) you are suggesting that we pull `TDyCreateDiagnostics` inside the dycore itself so that we can eliminate the diagnostics DM from the interface. I think that's fine, except that it then becomes harder to access the layout for the diagnostics fields, so the diagnostics vector becomes a "black box". Tell me what you think of this:

1. `TDyCreateDiagnosticsVector(TDy tdy, Vec *diag_vec)`: creates a global vector `diag_vec` with the proper layout for the dycore's diagnostics fields. It's a black box, as its layout is defined by the diagnostics DM, which isn't exposed to the user.
2.  `TDyCreateDiagnosticsSubVector(TDy tdy, PetscInt index, Vec *sub_vec)`: creates a global vector that stores a single diagnostic field (like liquid mass or saturation) with the given field index within the (hidden) diagnostics DM.
3. `TDyComputeDiagnostics(TDy tdy, Vec diag_vec)`: computes all diagnostic fields, storing the computed values in `diag_vec`. This is the same as what we have now, except we exclude the diagnostics DM from the interface.
4. `TDyGetSaturation(TDy tdy, Vec diag_vec, Vec sat_vec)`: extracts the saturation values from `diag_vec` into `sat_vec` (using e.g. `VecStride`).

Reply:
> It seems a little cumbersome to me to have all these DMs floating around

I agree that is why I think the DM and sub-DM should be internal to TDycore. These DMs can be created the first time user calls `TDyComputeDiagnostics`.

Regarding the proposed approach steps 1-5:

- It seems to me if I'm only interested in `TDyGetSaturation`, I have to also do steps 1-3. It is not a big deal for now as we don't have any user base, but it kinda becomes cumbersome.

Reply:
The reason I was proposing  `diag_vec` in addition to the liquid mass and saturation vectors is that we can compute all diagnostic fields at once, which is likely cheaper (especially on GPUs) than recomputing the saturation and liquid mass every time you ask for each of them. You should only have to do step 1 once in your program, and step 2 only once per diagnostic field to be extracted.

If this is confusing or cumbersome, we could consider some rules for how/when diagnostic fields are recomputed, and have the dycore store what we've been calling `diag_vec` internally. For example, we could say that diagnostic fields are updated every time a solver advances by a single time step. Or, we could have a function that explicitly updates all diagnostic fields internally so you can extract the most up-to-date values, e.g.

```
// Update all diagnostics fields given the current state of the dycore.
PetscErrorCode TDyUpdateDiagnostics(TDy tdy);
```

Then we would provide a functions for creating a Vec that can hold cell-centered scalar diagnostic fields (for example):

```
// Create a Vec that can store a cell-centered diagnostic field with a single component.
PetscErrorCode TDyCreateDiagnosticVector(TDy tdy, Vec *vec);
```

If we need vector-valued diagnostics, like the momentum vector within each cell, we could add an argument for the number of components in the diagnostic field. Finally, we can add functions for getting the saturation and liquid masses for each cell:

```
// Get the saturation in each cell, given a Vec created by TDyCreateDiagnosticVector.
PetscErrorCode TDyGetSaturation(TDy tdy, Vec sat_vec);
// Get the liquid mass contained in each cell, given a Vec created by TDyCreateDiagnosticVector.
PetscErrorCode TDyGetLiquidMass(TDy tdy, Vec liq_mass_vec);
```

So we would still have 4 functions, and the user still has to create the `Vec`s that store the saturation and liquid mass. But I don't think we could really simplify things any further than that, because the dycore doesn't know how many `Vec`s a user wants to use to store diagnostics.

How does this sound to you?

Reply:
@jeff-cohere I like your proposed approach of:
1. `TDyUpdateDiagnostics() // internal code`
2. `TDyCreateDiagnosticVector() // user code`
3. `TDyGetXYZ() // user code`